### PR TITLE
fix: 含文件附件的 bot 消息不参与折叠分组 (#322)

### LINKED ISSUE #322: Web端龙虾连续消息折叠时，含文件附件的消息不应被折叠
## 问题描述

龙虾在执行任务时经常连续输出多条消息（中间过程 + 最终交付物）。Web 端对连续消息有折叠展示机制（"显示更多"），**所有消息被一视同仁地折叠**，包括含文件附件的消息。

这导致龙虾产出的图片、PPT、PDF 等**关键交付物被折叠藏起来**，用户完全感知不到有文件产出，必须手动展开才能发现。

## 当前行为

- 龙虾连续输出 N 条消息 → 超过阈值后触发折叠
- 纯文字消息和含附件消息**被同等折叠**
- 用户看不到交付物的存在

## 期望行为

- **含文件附件的消息（图片/PPT/PDF/文档等）始终展示，不参与折叠**
- 纯文字的中间过程消息仍可正常折叠
- 折叠机制继续生效以避免刷屏，但交付物不被埋没

## 影响范围

- Web 端消息渲染 / 折叠逻辑
- iOS 端目前无此问题（无折叠机制）

## 优先级

高 — 直接影响用户获取龙虾工作成果的体验

### PR DESCRIPTION
**Pull request by pkuWMH**
## 问题

龙虾连续输出多条消息时，Web 端将所有消息折叠进 FoldSessionCard。含文件附件的消息（图片、PPT、PDF 等交付物）也被折叠，用户无法感知到有文件产出。

Closes #322

## 修改

在 `ConversationVM.buildRenderItems()` 中，含文件附件（image/gif/video/file/richText）的 bot 消息不再参与折叠分组：

- 新增 `hasFileAttachment()` 方法，判断消息是否包含文件类内容
- 含附件的 bot 消息作为折叠分组的边界：先 flush 当前分组，再作为独立消息渲染
- 纯文字的中间过程消息仍正常折叠

**示例：** bot 发送 [文字, 文字, 图片, 文字, 文字] → 渲染为 [折叠卡片(2条文字)] + [图片(始终可见)] + [折叠卡片(2条文字)]

## 测试

- 新增 4 个单元测试覆盖各种场景
- 已有 foldSessionSummary 测试全部通过（无 regression）

### CODE REVIEW

**Review by Jerry-Xin — DISMISSED** [review]
One-line summary: The PR is in scope and the folding change correctly keeps bot file-like deliverables visible while preserving text-only folding behavior.

💬 Non-blocking
- 🔵 Suggestion: `packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts:166` only covers `image` and `file`, while production excludes `gif`, `smallVideo`, and `richText` too at `packages/dmworkbase/src/Components/Conversation/vm.ts:291`. A small parameterized test over all excluded content types would reduce future enum drift risk.

✅ Highlights
- The attachment boundary logic in `packages/dmworkbase/src/Components/Conversation/vm.ts:399` is straightforward: it flushes the pending fold group, renders the attachment as a standalone message, and then starts fresh for later messages.
- The tests cover the important ordering behavior: standalone attachment, attachment as a boundary, and fold groups on both sides.

Verification caveat: I could not run the focused Vitest file because this checkout has no `node_modules`, and `pnpm exec vitest ...` failed with `Command "vitest" not found`.

**Review by lml2468 — CHANGES_REQUESTED** [review]
## Findings

### 🟡 P1: Tests do not cover the full behavior contract claimed by this PR

`packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts`

The new suite defines and asserts only `TEXT`, `IMAGE`, and `FILE`, and every test uses the default bot sender. The suite would still pass green if any of the following regressions were introduced:
- `gif` (`ContentType=3`) accidentally removed from `hasFileAttachment()`
- `smallVideo` (`ContentType=5`) accidentally removed
- `richText` (`ContentType=14`) accidentally removed
- The guard widened to user-sent file messages (which should still fold normally per the description: only bot deliverables are exempt)
- The flush call removed for the case where an attachment is the very first message in a run, or two attachment messages are adjacent

The implementation currently handles all of these correctly, but the regression net is narrower than the PR contract. Please extend the suite with at least:
- Table-driven cases asserting non-fold behavior for `gif=3`, `smallVideo=5`, `richText=14`
- A user-sent file message that **still folds** (or renders per existing user-message rules) — i.e. the guard is bot-only
- Boundary cases: attachment as first message; attachment as last message; two adjacent attachments with no text fold group between them

## Notes from inspection (no action needed)

- `hasFileAttachment()` covers the claimed file-bearing types: `image=2`, `gif=3`, `smallVideo=5`, `file=8`, `richText=14`. Constants verified on current base.
- RichText is reasonable to exclude from folding as a whole because the current RichText model is graphically mixed content and can contain embedded image/file blocks.
- The boundary semantics in `buildRenderItems()` are correct: for bot attachment messages, the pending fold group is flushed first, then the attachment message is pushed as a standalone item, preserving order and avoiding loss/duplication.
- The new branch only runs after `isBotMessage(message)`, so user messages do not enter the attachment-specific path.
- Revoked / sending messages remain excluded by `isBotMessage()` before the new branch; this change does not alter that behavior.
- Performance remains O(n); the added `switch` has no hot-path allocation.

## Recommendation

REQUEST_CHANGES — implementation is correct, but the test suite doesn't actually pin down the contract. Please broaden the tests; it's a small ask and consistent with the team standard that tests must turn red when the system breaks.

**Review by yujiawei — DISMISSED** [review]
# Code Review — PR #323 (octo-web)

## Verdict: APPROVED

This is a small, well-scoped fix (Closes #322): bot messages carrying file-like deliverables (image / gif / smallVideo / file / richText) are excluded from `FoldSessionCard` grouping, so users always see images, PDFs, PPTs etc. instead of having them hidden inside a collapsed card.

## Summary of changes

- `packages/dmworkbase/src/Components/Conversation/vm.ts` (+23): new private `hasFileAttachment()` and a guard inside `buildRenderItems()`'s bot-message branch — for an attachment message it flushes the pending fold group (`flushPendingSession(false)`) and renders the message standalone (`renderItems.push({ type: "message", message }); continue`).
- New test file `foldSessionFileAttachment.test.ts` (+238): 4 cases covering standalone image, standalone file, attachment breaking a run, and `foldSession + image + foldSession`.

## Verification

- ✅ **Logic is correct.** The guard sits inside the `isBotMessage` branch and runs before the time-window grouping, so an attachment always (a) terminates any in-progress fold group and (b) renders on its own, then grouping restarts cleanly for subsequent messages. The downstream `isStillActive` tail computation, the fold-session timer, the typing-merge pass, and the `afterFold` avatar marker are unaffected because an attachment leaves `pendingSessionMessages` empty at the boundary.
- ✅ **Content-type set matches Const.ts.** image(2) / gif(3) / smallVideo(5) / file(8) / richText(14) are correct per `Service/Const.ts`. voice(4) is deliberately left foldable (documented in the in-code comment). Non-deliverable types (card, location, mergeForward, lottie stickers) are correctly omitted; `screenshot(20)` is treated as a system/boundary message elsewhere in the codebase, so its omission is consistent with #322's stated scope.
- ✅ **Tests pass and exercise the real code path.** All 4 new tests run against a real `ConversationVM.buildRenderItems()` (not a reimplementation); the bot/group/timestamp setup correctly drives `isBotMessage` → `hasFileAttachment`. Locally `52/52` tests in the Conversation `__tests__` suite pass, no regressions. The mocked `MessageContentTypeConst` values match production exactly.

## Findings

No P0/P1 (merge-blocking) issues. The following are non-blocking suggestions:

### Suggestions (non-blocking)

- 🔵 **Test coverage gap (also raised in the prior review):** the 4 tests only assert `image` and `file`. `gif`, `smallVideo`, and `richText` are in the production switch but never asserted as message content — deleting one of those `case` lines would not fail any test. The switch is structurally uniform so the risk is low, but a parameterized case over all 5 excluded types plus one negative case (voice still folds) would lock the boundary against future enum drift. `foldSessionFileAttachment.test.ts:166-237`, `vm.ts:289-300`.
- 🔵 **Naming nuance:** `hasFileAttachment` returns true for `richText`, which can be pure formatted text with no embedded image. The effect is harmless (such a message simply renders un-folded), and including richText is a deliberate choice, but the method name is slightly broader than "attachment." `vm.ts:289-300`.
- 🔵 **Pre-existing test mock drift (not introduced by this PR):** the test mocks `ChannelTypeCommunityTopic: 6` while the real constant is `5`. No effect here (tests use a group channel; `hasFileAttachment` ignores channel type), but worth fixing if subregion-folding tests are later added to this file. `foldSessionFileAttachment.test.ts:25,126` vs `Service/Const.ts:136`.

Clean, focused fix that does exactly what #322 asks. Approving.

**Review by lml2468 — APPROVED** [review]
## Re-review of dea734d5145c (test: broaden fold session file attachment coverage)

Delta vs cdaaf009f104 is test-only (+107 -7 on the one test file). All prior P1 gaps are now closed:

✅ **Table-driven exclusion across all claimed types** — `it.each` covers `image=2`, `gif=3`, `smallVideo=5`, `file=8`, `richText=14`. If anyone removes a branch from `hasFileAttachment()`, the corresponding row turns red. (`foldSessionFileAttachment.test.ts:181-194`)

✅ **Voice still folds** — explicit negative test for `voice=4` (`foldSessionFileAttachment.test.ts:198-210`). Locks in the precise scope of the guard, not just the positive cases.

✅ **Bot-only guard** — `humanMessage(...)` helper added; user-sent file/image case asserts user messages are unaffected by the new path (`foldSessionFileAttachment.test.ts:214-226`).

✅ **Boundary coverage** — attachment as first message (`:230-245`), attachment as last message (`:249-264`), two adjacent attachments (`:268-280`). Any regression that forgot to flush the pending group, or duplicated the attachment item, or swapped order, would now fail.

The implementation in `vm.ts` was already correct in the previous round; this commit only broadens the regression net to match the PR's stated contract. No source changes.

## Recommendation

APPROVE

**Review by Jerry-Xin — APPROVED** [review]
Summary: The PR is in scope for `octo-web` and correctly keeps bot file-like attachment messages visible by making them boundaries in `ConversationVM.buildRenderItems()`.

💬 Non-blocking
- 🔵 Suggestion: [foldSessionFileAttachment.test.ts](/private/var/folders/fj/dr1p6h5n55l0ny89qk2n26tw0000gn/T/pr-review-209nb_7i/repo/packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts:166) duplicates content-type numbers in the test. This is acceptable with the current mocks, but importing or deriving the mocked constants would reduce drift risk if `MessageContentTypeConst` changes later.
- 🔵 Suggestion: [vm.ts](/private/var/folders/fj/dr1p6h5n55l0ny89qk2n26tw0000gn/T/pr-review-209nb_7i/repo/packages/dmworkbase/src/Components/Conversation/vm.ts:289) names the helper `hasFileAttachment`, while it also treats `richText` as a boundary. The behavior matches the PR description, but a name like `shouldRenderOutsideFoldSession` would be more semantically precise.

✅ Highlights
- The boundary logic is placed before normal bot-message grouping, so attachments flush any pending text group and render standalone as intended: [vm.ts](/private/var/folders/fj/dr1p6h5n55l0ny89qk2n26tw0000gn/T/pr-review-209nb_7i/repo/packages/dmworkbase/src/Components/Conversation/vm.ts:399).
- Tests cover first, last, adjacent, and between-text attachment positions, plus the explicit voice regression case: [foldSessionFileAttachment.test.ts](/private/var/folders/fj/dr1p6h5n55l0ny89qk2n26tw0000gn/T/pr-review-209nb_7i/repo/packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts:196).

Verification note: I attempted to run the targeted Vitest files with `pnpm exec vitest run ...`, but this checkout does not have a resolvable `vitest` binary installed, so I could not execute tests locally.

**Review by yujiawei — APPROVED** [review]
# Code Review — PR #323 (octo-web)

## Verdict: APPROVED

Small, well-scoped fix (Closes #322): bot messages carrying file-like deliverables (image / gif / smallVideo / file / richText) are excluded from `FoldSessionCard` grouping, so users always see images, PDFs, PPTs, etc. instead of having them buried inside a collapsed card. Plain-text intermediate bot messages still fold as before.

## Summary of changes

- `packages/dmworkbase/src/Components/Conversation/vm.ts` (+23): new private `hasFileAttachment()` plus a guard at the top of `buildRenderItems()`'s bot-message branch — for an attachment message it flushes the pending fold group (`flushPendingSession(false)`) and renders the message standalone (`renderItems.push({ type: "message", message }); continue`).
- New test `foldSessionFileAttachment.test.ts` (+338): table-driven exclusion across all 5 types, a voice-still-folds negative case, a bot-only-guard case, and boundary cases (first/last/adjacent attachments).

## Verification

- ✅ **Logic is correct.** The guard sits inside the `isBotMessage` branch and runs *before* the time-window grouping (`vm.ts:402-407`), so an attachment always (a) terminates any in-progress fold group and (b) renders on its own; grouping then restarts cleanly for subsequent messages. Because the attachment leaves `pendingSessionMessages` empty at the boundary, the trailing `isStillActive` tail computation, the fold-session timer, the typing-merge pass, and the `afterFold` avatar marker are all unaffected. No message loss, duplication, or ordering inversion.
- ✅ **Content-type set matches `Service/Const.ts`.** image(2) / gif(3) / smallVideo(5) / file(8) / richText(14) are correct (`Const.ts:49-62`). voice(4) is deliberately left foldable (documented in the in-code comment). Non-deliverable types (card, location, mergeForward, lottie stickers, screenshot) are correctly omitted, consistent with #322's scope.
- ✅ **Matches #322 intent exactly.** The issue asks for file-bearing deliverables to always show while plain-text intermediate messages still fold — which is precisely what the guard does.
- ✅ **Tests exercise the real code path** (`ConversationVM.buildRenderItems()`, not a reimplementation) and now pin the full contract: all 5 excluded types via `it.each`, voice as a negative case, the bot-only guard, and the first/last/adjacent boundary cases. The mocked `MessageContentTypeConst` values match production exactly.

## Findings

No P0/P1 (merge-blocking) issues. An adversarial multi-dimensional pass (correctness / tests / scope) surfaced only the following non-blocking nits:

### Nits (non-blocking)

- 🔵 **Mock drift on `ChannelTypeCommunityTopic`.** Both mocks declare `ChannelTypeCommunityTopic: 6` (`foldSessionFileAttachment.test.ts:25,126`) while the real constant is `5` (`Const.ts:136`). Harmless today — every test channel is a Group (`new Channel("g1", 2)`, `:148`), so the CommunityTopic branch of `supportsFolding` is never evaluated — but worth fixing to `5` so a future sub-topic test doesn't copy the wrong value.
- 🔵 **Fold feature also applies to sub-topics (CommunityTopic), but only the Group path is tested.** `supportsFolding` (`vm.ts:265-267`) covers both Group and CommunityTopic; the new behavior is fully exercised under Group, and the new code does not branch on channel type, so coverage is adequate — a parametrized case over both channel types would be a cheap hardening, not a requirement.
- 🔵 **`hasFileAttachment()` naming vs. richText behavior.** `richText` is unconditionally treated as un-foldable without inspecting whether its blocks actually contain an image/file (`vm.ts:289-300`). This is the PR's intended design (richText = 图文混排, listed as a deliverable, and the web client only emits richText when an image is present), and the worst case is purely cosmetic (a hypothetical text-only richText renders standalone instead of folding — no data loss). A name like `shouldRenderOutsideFoldSession` would read more precisely, but the behavior is correct.
- 🔵 **One redundant test.** The two-`humanMessage` case (`:233-245`) asserts human file/image messages render standalone — but human messages never fold regardless of this PR (they take the `else` branch at `vm.ts:419-420`), so it passes with or without the fix. It documents intent rather than guarding the new guard; a human-attachment-between-two-bot-texts case would be strictly stronger.

Clean, focused fix that does exactly what #322 asks. Approving.

**Review by mochashanyao — APPROVED** [review]
[Octo-Q · automated review]

Verdict: Approve — no blocking findings; notes below (data-flow traced).

---

# octo-web PR#323 Review Report
**Reviewer**: Octo-Q (automated review)
**Head SHA:** `0fb392e22823317d0e46a1493cf15db4faf6fa96`
**PR:** fix: 含文件附件的 bot 消息不参与折叠分组 (#322)

---

## 1. 验证结论

| 项目 | 状态 | 证据 |
|------|------|------|
| `hasFileAttachment()` 正确识别附件类型 | ✅ | `vm.ts:296-305` — switch 覆盖 image(2)/gif(3)/smallVideo(5)/file(8)/richText(14)，与 `Const.ts:51-62` 定义一致 |
| voice(4) 未被排除 | ✅ | `vm.ts:296-305` — voice 不在 switch case 中，仍可折叠 |
| 附件消息作为折叠边界正确 flush | ✅ | `vm.ts:407-412` — 先 `flushPendingSession(false)` 再 push standalone message |
| 仅对 bot 消息生效 | ✅ | `vm.ts:405` — `hasFileAttachment` 检查在 `isBotMessage(message)` 为 true 后才执行 |
| 用户发送的附件不受影响 | ✅ | `isBotMessage()` 在 `message.send=true` 时返回 false（`vm.ts:276`），human 消息走独立渲染路径 |
| 测试覆盖完整 | ✅ | 10 个测试用例覆盖所有排除类型、voice 仍折叠、用户消息不受影响、边界场景（首条/末条/相邻附件） |

## 2. 发现的问题

**无 P0/P1 问题。**

### P2 — 测试覆盖：未验证 `isActive` 状态

- **diff-scope:** new（新测试文件的覆盖范围选择）
- 测试仅验证了 render items 的结构（type/count/contentType），未验证 fold session 的 `isActive` 字段。当 bot 在 120s 活跃窗口内发送 text→image→text 时，image 前的 fold session 被 `flushPendingSession(false)` flush 为 inactive，但测试未断言此行为。
- **风险等级：** 低。`isActive` 逻辑与已有的 time-gap flush 一致（中间 flush 始终 `isActive=false`），且最终 flush 仍正确检测活跃状态。

### P2 — `richText` 分类为"附件"可能过宽

- **diff-scope:** new（本 PR 新增的 `hasFileAttachment` 方法）
- `richText`(14) 被归为"带附件的消息"。richText 是图文混排，可能包含纯格式化文本（无文件交付物）。但根据 PR 注释"这类消息是用户需要直接看到的交付物"以及 richText 通常包含嵌入图片/复杂排版的实际使用场景，此分类合理。
- **风险等级：** 低。若未来出现纯文本 richText 场景，可再细化。

## 3. 建议

- 可考虑补充一个测试用例验证 `isActive` 状态：bot 在 120s 内发送 text→image→text，断言 image 前的 fold session `isActive=false`、最终 fold session `isActive=true`。非阻塞。

## 4. 额外发现

无。

## 5. 数据流回溯

| 消费点 | 数据 | 上游来源 | 是否真的流到消费点 |
|--------|------|----------|-------------------|
| `hasFileAttachment(message)` | `message.contentType` | `MessageWrap.contentType` getter (`Model.tsx:355-356`) → `this.message.contentType` (WK SDK Message，服务端下发) | ✅ 直接透传，无中间转换 |
| `buildRenderItems` 循环 | `message` (附件) | `sourceMessages` 数组（过滤 typing 后的消息列表） | ✅ 附件消息进入 sourceMessages，在循环中被 `hasFileAttachment` 拦截 |
| `flushPendingSession(false)` | `pendingSessionMessages` | 循环中 `isBotMessage && !hasFileAttachment` 的 bot 消息累积 | ✅ 附件前的 bot 文本消息正确累积在 pendingSessionMessages 中 |
| `renderItems.push({type:"message", message})` | 附件 message | 循环当前迭代的 message | ✅ 附件消息不进入 pendingSessionMessages，直接 push 到 renderItems |

## 6. 三类盲点 checklist

- **C1 双路径 parity:** N/A — 本 PR 仅修改 `buildRenderItems` 中的 bot 消息处理路径，不涉及 add/remove、subscribe/unsubscribe 等对称操作。
- **C2 control-flow ordering / 嵌套复用:** Clear — `hasFileAttachment` 是纯函数（switch/return），无嵌套复用风险。`flushPendingSession` 的调用与已有 time-gap flush 路径一致，无双重作用。
- **C3 授权边界 ≠ 能力边界:** N/A — 纯前端 UI 渲染逻辑，不涉及权限/工具暴露。

---

**[Octo-Q] verdict: APPROVE** — 改动聚焦、逻辑正确、测试充分。无 P0/P1 问题。两个 P2 观察（isActive 测试覆盖、richText 分类）均为非阻塞建议。

File: packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+vi.mock("../../../Service/messageSelection", () => ({ isMessageSelectable: () => true }))
+
+import ConversationVM from "../vm"
+import { Channel } from "wukongimjssdk"
+
+const channel = new Channel("g1", 2)
+
+let seqCounter = 1
+
+function botMessage(contentType: number, timestamp: number, fromUID: string = "claude") {
```
Comment (lml2468) [inline packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts:152]: [minor] botMessage / humanMessage 是完全重复的实现,仅默认 fromUID 不同

botMessage(152-164) 与 humanMessage(174-186) 两个工厂函数的函数体逐字相同,唯一区别是默认参数 fromUID("claude" vs "alice")。这是明显的复制粘贴,增加阅读和后续维护成本(改一处要记得改两处)。建议抽成一个共享工厂,再用两个薄封装表达语义,例如:

```ts
function makeMessage(contentType: number, timestamp: number, fromUID: string) {
    const messageSeq = seqCounter++
    return { clientMsgNo: `msg-${messageSeq}`, messageSeq, fromUID, timestamp, contentType, revoke: false, send: false, from: { title: fromUID } } as any
}
const botMessage = (ct: number, ts: number, uid = "claude") => makeMessage(ct, ts, uid)
const humanMessage = (ct: number, ts: number, uid = "alice") => makeMessage(ct, ts, uid)
```

这样既保留两个有语义的入口(bot/human),又消除重复。

File: packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        fromUID,
+        timestamp,
+        contentType,
+        revoke: false,
+        send: false,
+        from: { title: fromUID },
+    } as any
+}
+
+const TEXT = 1
```
Comment (lml2468) [inline packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts:166]: [nit] 内容类型常量声明被夹在两个 helper 函数之间,打断阅读顺序

TEXT/IMAGE/GIF/... 等常量(166-172)被插在 botMessage(152) 和 humanMessage(174) 两个工厂函数中间,使两个本应相邻、对照阅读的同类 helper 被割裂,常量本身也显得位置随意。建议把这组 contentType 常量上移到 seqCounter 之前(或紧跟 import 后的顶部常量区),让两个 message 工厂连续排布,结构更清晰。

**Review by lml2468 — COMMENTED** [review]
## PR #323 Review 结论(comment-only)

**改动概览**
2 文件 +361/-0,关联 #322。新增折叠会话文件附件可见性的单元测试,并在 `vm.ts` 中加入「含文件附件的 bot 消息不参与折叠分组」的守卫逻辑。

**测试验证(动态)**
- testsPass = ✅ true
- 无新增失败 = ✅ true(基线 0 → 0)
- build = ✅ true
- lint = ✅ true

**结论**
- 阻断问题:**0** 条
- 小问题 / nit:**2** 条(均为测试文件可读性建议,见行内评论:`foldSessionFileAttachment.test.ts:152`、`:166`)
- 已丢弃:**6** 条经核验不成立的误报

整体改动正确、测试充分、构建与 lint 通过,无阻断项。剩余两条为 test 文件中的可读性/重复消除建议,不影响合并。

**表态:comment-only** —— 此为边界情况,不主动 approve / request-changes,留人工裁决。

**Review by lml2468 — APPROVED** [review]
## ✅ APPROVE — #323 `0fb392e2`(基于最新 commit 重审)

跑了 pr-review workflow(comment-only / 0 blocker,丢弃 6 误报,测试+build+lint 全绿)并独立核实折叠边界逻辑。修对了 #322。

### ✅ 核心逻辑已核实(vm.ts)
带附件的 bot 消息(image/gif/smallVideo/file/richText)作为折叠边界:`hasFileAttachment` 命中 → `flushPendingSession(false)` 先冲掉已累积分组 → 独立 `renderItems.push` → `continue`。顺序正确(先 flush 再独立渲染,保证交付物在折叠卡外始终可见)。**voice=4 特意排除**仍可折叠,语义正确。守卫在 `isBotMessage` 内,不影响用户消息。

### ✅ 测试覆盖到位
7 个用例覆盖关键边界:voice 仍折叠、bot-only 守卫、首/末/相邻附件独立、图片打断分组、图片两侧两段 text 各自折叠 — 正是这次改动的全部分支。

### 🟡 Fast-follow(非阻塞,均在测试文件)
1. `foldSessionFileAttachment.test.ts:152` — `botMessage` 与 `humanMessage` 两工厂函数体逐字相同,仅默认 `fromUID` 不同(claude/alice)。建议抽共享 `makeMessage` + 两个薄封装,消除复制粘贴。
2. `:166` — `TEXT/IMAGE/...` 常量夹在两个工厂函数之间,割裂同类 helper。建议上移到顶部常量区。

0 blocker、逻辑正确、CI 全绿 → APPROVE。(注:merge 仍 BLOCKED 是系统性 check-sprint 看板门,非代码问题,需 maintainer 处理。)

**Review by yujiawei — APPROVED** [review]
# Code Review — PR #323 (octo-web)

Reviewed at head `0fb392e22823317d0e46a1493cf15db4faf6fa96` (base `main`, merge-base current). Scope: `packages/dmworkbase/src/Components/Conversation/vm.ts` (+23) and a new test file (+338).

## Summary

The change does what #322 needs: bot messages carrying a deliverable (image / gif / smallVideo / file / richText) now act as a fold-group boundary — the pending fold session is flushed, the attachment renders standalone, and pure-text process messages on either side still fold. The implementation is small, the boundary handling at first/last/adjacent positions is correct, and the test coverage is solid (excluded types, voice-still-folds, human-message guard, all four boundary positions, plus the original structural cases).

`hasFileAttachment` is correctly scoped to bot messages only (it is called inside the `isBotMessage` branch), so human-sent files are unaffected — they already render standalone. Voice (type 4) is deliberately excluded and stays foldable, which matches the stated intent.

Verdict: **Approve.** No correctness, security, data-loss, or build-break issues. Two advisory (P2) notes below worth a follow-up but not blocking.

## Findings

### P2 — `richText` (type 14) is treated as an attachment unconditionally

`vm.ts:301` adds `MessageContentTypeConst.richText` to `hasFileAttachment()` with no inspection of the block payload. But richText is 图文混排 (mixed text/image/file blocks) and can be **text-only** — `RichTextContent.ts:140` decodes a plain raw string into a single `{ type: "text" }` block. A bot that sends consecutive text-only richText replies within the 120s window will therefore no longer fold them into a FoldSessionCard; each renders standalone, and any active-session / typing preview that would have attached to that fold session is lost for that case.

This fails safe (toward visibility, which is the PR's whole point), so it is not a blocker. But if bots emit text-only richText for ordinary process chatter, the fold feature silently stops applying to it. Consider inspecting the blocks and returning `true` only when an `image`/`file` block is present (e.g. `content.some(b => b.type === "image" || b.type === "file")`), with `text`-only richText falling through to normal folding.

### P2 — Allowlist may miss other non-text deliverables

`hasFileAttachment()` uses a strict allowlist. Other content types a bot could plausibly send as a deliverable — `mergeForward` (11, 合并转发), `card` (7), `location` (6) — are not covered and would still be folded/hidden inside the card. This is a judgment call: an allowlist is precise and avoids over-exposing chatter, but it requires maintenance as new deliverable types appear. If "any non-text deliverable should stay visible" is the real requirement, a blocklist (fold only `text` and `voice`, surface everything else) would fail open instead. Not blocking for the image/file/PDF cases #322 targets.

## Notes (non-blocking)

- An adversarial check raised whether the `continue` at `vm.ts:411` could skip tail-of-loop bookkeeping (time dividers / avatar grouping). Verified it does not: the loop body has no per-iteration tail state — the existing non-attachment bot branch also ends in `continue`, and post-fold avatar correction runs after the loop via `afterFoldSessionClientMsgNos`. No issue here.
- Tests are well structured but run against hand-rolled message stubs and a fully mocked SDK rather than real `MessageWrap` instances. That is reasonable for a VM unit test; just be aware the richText case above is not exercised because the stubs only carry a `contentType`, not block content.

File: packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts (hunk 1 above)
Comment (Jerry-Xin) [reply packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts:152]: Good catch. Agreed this is a copy-paste duplication. Since this is a test-only readability nit and not behavioral, we'll fold the shared-factory refactor into a follow-up cleanup rather than reopen this merge-ready PR. Tracking it as a follow-up.

File: packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts (hunk 2 above)
Comment (Jerry-Xin) [reply packages/dmworkbase/src/Components/Conversation/__tests__/foldSessionFileAttachment.test.ts:166]: Agreed, the constant placement could read better. Same as above: a test-file ordering nit, will be addressed in the follow-up cleanup. Resolving for now.
